homeassistant/components/matter/water_heater.py

```diff
@@ -0,0 +1,205 @@
+
+    @callback
+    def _update_from_device(self) -> None:
+        """Update from device."""
+        # self._calculate_features()
+        self._attr_current_temperature = self._get_temperature_in_degrees(
+            clusters.Thermostat.Attributes.LocalTemperature
+        )
+        self._attr_target_temperature = self._get_temperature_in_degrees(
+            clusters.Thermostat.Attributes.OccupiedHeatingSetpoint
+        )
+        BoostState = self.get_matter_attribute_value(
+            clusters.WaterHeaterManagement.Attributes.BoostState
+        )
+        if (BoostState) == clusters.WaterHeaterManagement.Enums.BoostStateEnum.kActive:
+            self._attr_current_operation = STATE_HIGH_DEMAND
+        else:
+            self._attr_current_operation = STATE_ECO
+        self._attr_temperature = cast(
+            float,
+            self._get_temperature_in_degrees(
+                clusters.Thermostat.Attributes.OccupiedHeatingSetpoint
+            ),
+        )
+        self._attr_min_temp = cast(
+            float,
+            self._get_temperature_in_degrees(
+                clusters.Thermostat.Attributes.AbsMinHeatSetpointLimit
+            ),
+        )
+        self._attr_max_temp = cast(
+            float,
+            self._get_temperature_in_degrees(
+                clusters.Thermostat.Attributes.AbsMaxHeatSetpointLimit
+            ),
+        )
+        self._attr_target_temperature_low = self._attr_min_temp
+        self._attr_target_temperature_high = self._attr_max_temp
+
+    @callback
+    def _get_temperature_in_degrees(
+        self, attribute: type[clusters.ClusterAttributeDescriptor]
+    ) -> float | None:
+        """Return the scaled temperature value for the given attribute."""
+        if value := self.get_matter_attribute_value(attribute):
+            return float(value) / TEMPERATURE_SCALING_FACTOR
+        return None
+
+
+# Discovery schema(s) to map Matter Attributes to HA entities
+DISCOVERY_SCHEMAS = [
+    MatterDiscoverySchema(
+        platform=Platform.WATER_HEATER,
+        entity_description=WaterHeaterEntityDescription(
+            key="MatterWaterHeater",
+        ),
+        entity_class=MatterWaterHeater,
+        required_attributes=(
+            clusters.Thermostat.Attributes.OccupiedHeatingSetpoint,
+            clusters.Thermostat.Attributes.AbsMinHeatSetpointLimit,
+            clusters.Thermostat.Attributes.AbsMaxHeatSetpointLimit,
+            clusters.Thermostat.Attributes.LocalTemperature,
+            clusters.WaterHeaterManagement.Attributes.FeatureMap,
+        ),
+        optional_attributes=(
+            clusters.WaterHeaterManagement.Attributes.HeaterTypes,
+            clusters.WaterHeaterManagement.Attributes.BoostState,
+            clusters.WaterHeaterManagement.Attributes.HeatDemand,
+        ),
+        device_type=(device_types.WaterHeater,),
+        allow_multi=True,  # also used for sensor entity
+    ),
```

Comment:
We probably have to set name to `None` to make the device name be the name of the entity, as I am assuming the water heater always is the main entity of the device